niworkflows/utils/spaces.py

```diff
@@ -0,0 +1,519 @@
+    def spaces(self, value):
+        self._spaces = value
+
+    def add(self, value):
+        """Add one more space, without erroring if it exists."""
+        if value not in self:
+            self.spaces += [self.check_space(value)]
+
+    def append(self, value):
+        """Concatenate one more space."""
+        if value not in self:
+            self.spaces += [self.check_space(value)]
+            return
+
+        raise ValueError('space "%s" already in spaces.' % str(value))
+
+    def insert(self, index, value, error=True):
+        """Concatenate one more space."""
+        if value not in self:
+            self.spaces.insert(index, self.check_space(value))
+        elif error is True:
+            raise ValueError('space "%s" already in spaces.' % str(value))
+
+    def get_std_spaces(self, only_names=True, dim=(2, 3)):
+        """Return only standard spaces."""
+        names = []
+        std_spaces = []
+        for s in self._spaces:
+            if s.standard and s.dim in dim and s.fullname not in names:
+                names.append(s.fullname)
+                std_spaces.append(s)
+        if only_names:
+            return names
+        return std_spaces
+
+    def get_templates(self, dim=(2, 3)):
+        """Return output spaces."""
+        return [s for s in self._spaces
+                if s.standard and hasspec(s.spec, ('res', 'den')) and s.dim in dim]
+
+    def get_nonstd_spaces(self, only_names=True, dim=(2, 3)):
+        """Return nonstandard spaces."""
+        return [s.name if only_names else s
+                for s in self._spaces
+                if not s.standard and s.dim in dim]
+
+    def get_fs_spaces(self):
+        """Return FreeSurfer spaces."""
+        return [s.legacyname or s.name
+                for s in self._spaces
+                if s.name in ("fsaverage", "fsnative")]
+
+
+def hasspec(value, specs):
+    """Check whether any of the keys are in a dict."""
+    for s in specs:
+        if s in value:
+            return True
+    return False
+
+
+def _expand_entities(entities):
+    """
+    Generate multiple replacement queries based on all combinations of values.
+
+    Ported from PyBIDS
+
+    Examples
+    --------
+    >>> entities = {'subject': ['01', '02'], 'session': ['1', '2'], 'task': ['rest', 'finger']}
+    >>> out = _expand_entities(entities)
+    >>> len(out)
+    8
+    >>> {'subject': '01', 'session': '1', 'task': 'rest'} in out
+    True
+    >>> {'subject': '02', 'session': '1', 'task': 'rest'} in out
+    True
+    >>> {'subject': '01', 'session': '2', 'task': 'rest'} in out
+    True
+    >>> {'subject': '02', 'session': '2', 'task': 'rest'} in out
+    True
+    >>> {'subject': '01', 'session': '1', 'task': 'finger'} in out
+    True
+    >>> {'subject': '02', 'session': '1', 'task': 'finger'} in out
+    True
+    >>> {'subject': '01', 'session': '2', 'task': 'finger'} in out
+    True
+    >>> {'subject': '02', 'session': '2', 'task': 'finger'} in out
+    True
```

Comment:
This is deterministic in Python 3.6+, too, right? Can we do the `PY_VERSION` trick again?

Reply:
done!